src/dbg/advancedanalysis.cpp

```diff
@@ -0,0 +1,373 @@
+#include "advancedanalysis.h"
+#include <queue>
+#include "console.h"
+#include "filehelper.h"
+#include "function.h"
+#include "xrefs.h"
+#include "encodemap.h"
+
+AdvancedAnalysis::AdvancedAnalysis(duint base, duint size, bool dump)
+    : Analysis(base, size),
+      mDump(dump)
+{
+    mEncMap = new byte[size];
+    memset(mEncMap, 0, size);
+}
+
+void AdvancedAnalysis::Analyse()
+{
+    linearXrefPass();
+    findEntryPoints();
+    analyzeCandidateFunctions(true);
+    findFuzzyEntryPoints();
+    analyzeCandidateFunctions(true);
+    findInvalidXrefs();
+    writeDataXrefs();
+}
+
+void AdvancedAnalysis::SetMarkers()
+{
+    if(mDump)
+        for(const auto & function : mFunctions)
+            FileHelper::WriteAllText(StringUtils::sprintf("cfgraph_" fhex ".dot", function.entryPoint), function.ToDot());
+
+    byte* buffer = (byte*)EncodeMapGetBuffer(mBase, true);
+    memcpy(buffer, mEncMap, mSize);
+    EncodeMapReleaseBuffer(buffer);
+
+    XrefDelRange(mBase, mBase + mSize - 1);
+    for(const auto & vec : mXrefs)
+    {
+        for(const auto & xref : vec.second)
+        {
+            if(xref.valid)
+                XrefAdd(xref.addr, xref.from);
+        }
+    }
+
+    FunctionClear();
+    for(const auto & function : mFunctions)
+    {
+        duint start = ~0;
+        duint end = 0;
+        duint icount = 0;
+        for(const auto & node : function.nodes)
+        {
+            icount += node.second.icount;
+            start = min(node.second.start, start);
+            end = max(node.second.end, end);
+        }
+        if(!FunctionAdd(start, end, false, icount))
+        {
+            FunctionDelete(start);
+            FunctionDelete(end);
+            FunctionAdd(start, end, false, icount);
+        }
+    }
+    GuiUpdateAllViews();
+}
+
+void AdvancedAnalysis::analyzeFunction(duint entryPoint, bool writedata)
+{
+    //BFS through the disassembly starting at entryPoint
```

Comment:
I know, maybe I'll resolve overlapping problem later. There could be gaps between basic blocks, especially in control flow obfuscated executable.